hw/xbox/xid-sdl.c

```diff
@@ -464,6 +446,81 @@ static void usb_xid_class_initfn(ObjectClass *klass, void *data)
     uc->handle_attach  = usb_desc_attach;
 }
 
+static int apply_haptic_effect(USBXIDState *s, SDL_HapticEffect *effect) {
+    int effect_id = SDL_HapticNewEffect(s->sdl_haptic, effect);
+    if (effect_id == -1) {
+        const char* error = SDL_GetError();
+        fprintf(stderr, "Failed to create haptic effect! SDL Error: %s\n", error);
+        SDL_HapticClose(s->sdl_haptic);
+        s->sdl_haptic = NULL;
+        s->sdl_haptic_effect = NULL;
+        return -1;
+    }
+
+    int run_result = SDL_HapticRunEffect(s->sdl_haptic, effect_id, 1);
+    if (run_result == -1) {
+        const char* error = SDL_GetError();
+        fprintf(stderr, "Failed to run haptic effect! SDL Error: %s\n", error);
+        SDL_HapticClose(s->sdl_haptic);
+        s->sdl_haptic = NULL;
+        s->sdl_haptic_effect = NULL;
+        return -1;
+    }
+
+    s->sdl_haptic_effect = effect;
+    s->sdl_haptic_effect_id = effect_id;
+    return 0;
+}
+
+static int haptic_effect_init(USBXIDState *s, SDL_HapticEffect *effect) {
+    if (SDL_HapticQuery(s->sdl_haptic) & SDL_HAPTIC_LEFTRIGHT) {
+        effect->leftright = (SDL_HapticLeftRight){
+            .type = SDL_HAPTIC_LEFTRIGHT,
+            .length = SDL_HAPTIC_INFINITY,
+            .large_magnitude = 0x0000,
+            .small_magnitude = 0x0000,
+        };
+
+        return apply_haptic_effect(s, effect);
```

Comment:
As your current `apply_haptic_effect` creates a new effect, I'd advise to do `apply_haptic_effect` in the caller (which knows `s` and `effect` anyway). It would make this function *much* more reusable as it could be used to construct an haptic effect for either `SDL_HapticCreateEffect` or `SDL_HapticUpdateEffect`.